### PR TITLE
To Do List features

### PR DESCRIPTION
**Pull request by Blaise-Shyaka**
In this project, we have done the following:
- Created factory functions to handle the creation of project and task instances
- Created the user interface
- Generated DOM elements using JavaScript
- Added DOM event handlers

### CODE REVIEW

File: src/scripts/components/headerComponent.js [hunk 1]
```diff
@@ -0,0 +1,20 @@
+import setAttributes from '../setAttributes';
+
+export default function createHeader() {
```
Comment (addod19) [inline src/scripts/components/headerComponent.js]: Kindly use ES6 arrow function instead

File: src/scripts/components/modalComponent.js [hunk 2]
```diff
@@ -0,0 +1,123 @@
+import setAttributes from '../setAttributes';
+
+export function addProjectModal() {
```
Comment (addod19) [inline src/scripts/components/modalComponent.js]: Kindly use ES6 arrow function instead

File: src/scripts/components/modalComponent.js [hunk 3]
```diff
@@ -0,0 +1,123 @@
...
+  modalContent.append(modalHeader, modalBody, modalFooter);
+
+  modalDialog.appendChild(modalContent);
+
+  wrapperModal.appendChild(modalDialog);
+
+  return wrapperModal;
+}
+
+export function deleteProjectModal() {
```
Comment (addod19) [inline src/scripts/components/modalComponent.js]: Kindly use ES6 arrow function instead

File: src/scripts/components/projectComponent.js [hunk 4]
```diff
@@ -0,0 +1,87 @@
+import setAttributes from '../setAttributes';
+import displayTask from './taskComponent';
+
+function allProjectTasks(projectId, tasksList) {
```
Comment (addod19) [inline src/scripts/components/projectComponent.js]: Kindly use ES6 arrow function instead

File: src/scripts/components/projectComponent.js [hunk 5]
```diff
@@ -0,0 +1,87 @@
...
+      title,
+      description,
+      date,
+      priority);
+    tasksContainer.appendChild(markup);
+  });
+  return tasksContainer;
+}
+
+export default function displayProject(projectId, project) {
```
Comment (addod19) [inline src/scripts/components/projectComponent.js]: Kindly use ES6 arrow function instead

File: src/scripts/components/taskComponent.js [hunk 6]
```diff
@@ -0,0 +1,103 @@
+import setAttributes from '../setAttributes';
+
+export default function displayTask(
```
Comment (addod19) [inline src/scripts/components/taskComponent.js]: Kindly use ES6 arrow function instead

File: src/scripts/components/taskModalComponent.js [hunk 7]
```diff
@@ -0,0 +1,164 @@
+import setAttributes from '../setAttributes';
+
+export function addTaskModal() {
```
Comment (addod19) [inline src/scripts/components/taskModalComponent.js]: Kindly use ES6 arrow function instead

File: src/scripts/components/taskModalComponent.js [hunk 8]
```diff
@@ -0,0 +1,164 @@
...
+
+  const selectTask = document.createElement('select');
+  setAttributes(selectTask, {
+    class: 'form-control mb-3', id: 'task-select', 'aria-label': 'Choose your priority',
+  });
+
+  // const optionMessage = document.createElement('option');
+  // setAttributes(optionMessage, {
+  //   selected: 'true'
+  // });
```
Comment (addod19) [inline src/scripts/components/taskModalComponent.js]: Kindly remove commented code :smile:

File: src/scripts/components/taskModalComponent.js [hunk 9]
```diff
@@ -0,0 +1,164 @@
...
+  modalContent.append(modalHeader, modalBody, modalFooter);
+
+  modalDialog.appendChild(modalContent);
+
+  wrapperModal.appendChild(modalDialog);
+
+  return wrapperModal;
+}
+
+export function deleteTaskModal() {
```
Comment (addod19) [inline src/scripts/components/taskModalComponent.js]: Kindly use ES6 arrow function instead

File: src/scripts/dom-handlers/addProject.js [hunk 10]
```diff
@@ -0,0 +1,52 @@
+/* eslint-disable no-undef */
+/* eslint-disable import/no-cycle */
+import newProject from '../project';
+import allProjectsMarkup from './renderProjects';
+
+const projects = [newProject('Default Project')];
+
+function cleanForm(elts = []) {
```
Comment (addod19) [inline src/scripts/dom-handlers/addProject.js]: Kindly use ES6 arrow function instead

File: src/scripts/dom-handlers/addProject.js [hunk 11]
```diff
@@ -0,0 +1,52 @@
...
+  elts.forEach(elt => {
+    elt.value = '';
+  });
+}
+
+function handleGeneratingNewInstance() {
```
Comment (addod19) [inline src/scripts/dom-handlers/addProject.js]: Kindly use ES6 arrow function instead

File: src/scripts/dom-handlers/addProject.js [hunk 12]
```diff
@@ -0,0 +1,52 @@
...
+    modal.hide();
+    allProjectsMarkup(projects);
+    // editProject();
+    // render();
+  } else {
+    projectTitleInput.className = 'form-control is-invalid was-validated';
+  }
+}
+
+function handleAddingProject() {
```
Comment (addod19) [inline src/scripts/dom-handlers/addProject.js]: Kindly use ES6 arrow function instead

File: src/scripts/dom-handlers/addProject.js [hunk 13]
```diff
@@ -0,0 +1,52 @@
...
+  const newTitle = document.querySelector('#project-title-input');
+  newTitle.value = '';
+
+  const header = document.querySelector('#exMl');
+  header.innerHTML = 'Add Project';
+  const saveChangesBtn = document.querySelector('#save-btn');
+  saveChangesBtn.addEventListener('click', handleGeneratingNewInstance);
+}
+
+function addProject() {
```
Comment (addod19) [inline src/scripts/dom-handlers/addProject.js]: Kindly use ES6 arrow function instead

File: src/scripts/dom-handlers/addTask.js [hunk 14]
```diff
@@ -0,0 +1,65 @@
+/* eslint-disable no-undef */
+/* eslint-disable import/no-cycle */
+import { projects, cleanForm } from './addProject';
+import allProjectsMarkup from './renderProjects';
+import setAttributes from '../setAttributes';
+
+function handleAddingTask(event) {
```
Comment (addod19) [inline src/scripts/dom-handlers/addTask.js]: Kindly use ES6 arrow function instead

File: src/scripts/dom-handlers/addTask.js [hunk 15]
```diff
@@ -0,0 +1,65 @@
...
+}
+
+function taskPriority(priorityLevel) {
+  if (priorityLevel === '1') { return 'Low'; }
+  if (priorityLevel === '2') { return 'Medium'; }
+  if (priorityLevel === '3') { return 'High'; }
+  return priorityLevel;
+}
+
+function handleAdd() {
```
Comment (addod19) [inline src/scripts/dom-handlers/addTask.js]: Kindly use ES6 arrow function instead

File: src/scripts/dom-handlers/deleteProject.js [hunk 16]
```diff
@@ -0,0 +1,41 @@
+/* eslint-disable no-undef */
+/* eslint-disable import/no-cycle */
+import { projects } from './addProject';
+import allProjectsMarkup from './renderProjects';
+import setAttributes from '../setAttributes';
+
+function handleDeletingProject(event) {
```
Comment (addod19) [inline src/scripts/dom-handlers/deleteProject.js]: Kindly use ES6 arrow function instead

File: src/scripts/dom-handlers/deleteProject.js [hunk 17]
```diff
@@ -0,0 +1,41 @@
...
+}
+
+export default function deleteProject() {
+  const deleteProjectBtn = document.querySelectorAll('.deleteProject');
+  deleteProjectBtn.forEach(delPrjct => {
+    delPrjct.addEventListener('click', handleDeletingProject);
+  });
+}
+
+export function handleDeleteProject() {
```
Comment (addod19) [inline src/scripts/dom-handlers/deleteProject.js]: Kindly use ES6 arrow function instead

File: src/scripts/dom-handlers/deleteTask.js [hunk 18]
```diff
@@ -0,0 +1,45 @@
+/* eslint-disable no-undef */
+/* eslint-disable import/no-cycle */
+import { projects } from './addProject';
+import allProjectsMarkup from './renderProjects';
+import setAttributes from '../setAttributes';
+
+function handleDeletingTask(event) {
```
Comment (addod19) [inline src/scripts/dom-handlers/deleteTask.js]: Kindly use ES6 arrow function instead

File: src/scripts/dom-handlers/deleteTask.js [hunk 19]
```diff
@@ -0,0 +1,45 @@
...
+
+  title.innerHTML = `Do you want to delete "${projects[correspondPrjctIndex].tasksList[correspondingIndex].title}"?`;
+
+  const deleteBtn = document.querySelector('#delBtnTask');
+  setAttributes(deleteBtn, {
+    'data-index': correspondingIndex, 'data-index-project': correspondPrjctIndex,
+  });
+}
+
+export default function deleteTask() {
```
Comment (addod19) [inline src/scripts/dom-handlers/deleteTask.js]: Kindly use ES6 arrow function instead

File: package.json [hunk 20]
```diff
@@ -0,0 +1,42 @@
...
+    "eslint-plugin-import": "^2.20.2",
+    "file-loader": "^6.2.0",
+    "image-webpack-loader": "^7.0.1",
+    "style-loader": "^2.0.0",
+    "stylelint": "13.3.x",
+    "stylelint-config-standard": "20.0.x",
+    "stylelint-csstree-validator": "^1.9.0",
+    "stylelint-scss": "3.17.x",
+    "webpack": "^5.21.2",
+    "webpack-cli": "^4.5.0"
```
Comment (addod19) [inline package.json]: Kindly add `webpack-dev-server` to our ppackage.json file :muscle:

File: package.json [hunk 21]
```diff
@@ -0,0 +1,42 @@
+{
+  "name": "to-do-list",
+  "version": "1.0.0",
+  "description": "",
+  "main": "index.js",
+  "scripts": {
+    "test": "echo \"Error: no test specified\" && exit 1",
+    "build": "webpack",
+    "watch": "webpack --watch"
+  },
```
Comment (addod19) [inline package.json]: Let us replace our script from 
````
 "scripts": {
    "test": "echo \"Error: no test specified\" && exit 1",
    "build": "webpack",
    "watch": "webpack --watch"
  }

 to 

"scripts": {
    "dev": "webpack --mode development",
    "build": "webpack --mode production",
    "start": "webpack-dev-server --mode development --open" // optional
  },

**Review by addod19 — CHANGES_REQUESTED** [review]
## Status: Require changes :recycle: :traffic_light: 

Hi @Blaise-Shyaka and @guillainbisimwa, great job on this project :muscle: The App looks great, the functionality works as expected, we can delete, add, edit a task, we can do the same to the project as well :smile: Awesome. The README file is very professional with a working live link :100: The PR has a meaningful title and a short description :fireworks: Commit messages are great and easy to understand :1st_place_medal: Great implementation of the single responsibility design pattern :slightly_smiling_face: 
However, we have a few comments we need to address before proceeding, kindly check the comments below:

- Kindly implement localStorage so the data remains in the browser even when we refresh the browser. Let us implement this :+1: 
- Kindly add webpack-dev-server to our package.json file
- Kindly use ES6 arrow function

Once these are resolved, kindly submit another code review request :smile: 
_Happy coding_ :man_technologist: 
[Daniel Larbi Addo](https://www.linkedin.com/in/daniel-larbi-addo/)

**Review by NtwaliHeritier — CHANGES_REQUESTED** [review]
### **STATUS: CHANGES REQUESTED🛑🛑🛑**

Hello team,
Great work on this project. The UI is 👌 and the app runs smoothly.
However small changes are required to make it even more outstanding.

[REQUIRED]: Kindly save data in the localStorage. This will prevent the loss of data after refreshing a page.

Make the requested changes and ask for another review👌👌👌
Happy coding 💻

**Review by mikenath223 — APPROVED** [review]
### CONGRATULATIONS  👏👏👏

This project meets the requirements for approval.

**STATUS: APPROVED ✅**


Don't forget to fill the project completion form👍
Good luck with the next project.

**HAPPY CODING 🎉💻🎉**

**Michgolden Ukeje (CR)**

**Pull request merged**